### DISCUSSION TITLE
On template props: meaning and data structure

### DISCUSSION THREAD

**Discussion by sushi2all** [pdfme/pdfme#267] [Q&A]
Hi all, first of all shout out to @hand-dot for jumpstarting a revolution in web-based pdf management. Pdfme looks amazing and is increasing in popularity by the day, so thank you! 🙏🏻

I have some question about Template props:

- I understand the `columns` prop lists all the fields that are present in the whole pdf. However they are also mapped in the schemas array, and `columns` is not required to render a pdf. So what is it for?
- The `sampledata` array contains a single object with each field name as key and placeholder text as value. Why is it an array if all fields are in a single object?

This is just out of curiosity to make sure I'm using the library in the best way possible.

Thanks in advance for any insight!

**Comment by hand-dot** [comment]
@sushi2all 

Sorry for the late reply.

I'm using 'columns' to preserve the order of the fields. In the form's tabindex and in the service I'm running, input can be done in Excel format, so the order of the table columns is determined by 'columns'. However, the latest JS objects guarantee the order of the keys, so it may be possible to remove them after testing in each browser.

'sampledata' is defined as an array, similar to the inputs of the generator.
